Move GenerationMixin to separate file

yjernite opened the pull request:
This PR splits the `modeling_utils.py` and `modeling_tf_utils.py` by moving the code and methods related to generation to `modeling_generation.py` and `modeling_tf_generation.py` respectively.

Both of these files were getting pretty long, with the code dedicated to generation taking about 1000 LOC in each while being completely disjoint from the rest. This re-organization should make the code easier to read and contribute to.

There are no functional changes, I literally just created a new Mixin class for each and moved the functions as is.

sshleifer commented on src/transformers/modeling_generation.py:
```diff
@@ -0,0 +1,989 @@
+logger = logging.getLogger(__name__)
+
+
+class ModuleGenerationMixin:
```
(nit) naming it `ModuleGenerationMixin` makes it seem like it will work for any `nn.Module`, like `ModuleUtilsMixin` can. 

I propose `GenerationMixin`/`TFGenerationMixin`

sshleifer commented on src/transformers/modeling_generation.py:
```diff
@@ -0,0 +1,989 @@
+
+class ModuleGenerationMixin:
+    """
+    A class contraining all of the functions supporting generation, to be used as a mixin.
```
you can say explicitly that the only child is `PretrainedModel`

patrickvonplaten commented:
Awesome, very much in favor of this change!

patrickvonplaten commented on src/transformers/__init__.py:
```diff
@@ -406,9 +407,11 @@
 if is_tf_available():
-    from .modeling_tf_utils import (
+    from .modeling_tf_generation import (
         shape_list,
```
Why is `shape_list` in `modeling_tf_generation`?  I think we should leave it in `modeling_tf_utils` and import it. I think it's heavily used in other files as well

patrickvonplaten replied:
We could import it dynamically if problems with circular imports arrive

yjernite replied:
yeah I was having trouble with circular imports, will keep trying to make it work :)

thomwolf replied:
agree with @patrickvonplaten here. It's a very simple utility so don't fight too much with imports and copy it both in `modeling_tf_utils` and `modeling_tf_generation_utils` if you want (cf the [Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) that we follow in the lib, duplicating once is ok.

yjernite commented:
I changed the name and mentioned the intended child class in the comment as suggested by @sshleifer 

I tried playing around with `importlib` to import `shape_list` dynamically from `modeling_tf_utils.py` but I'm getting stumped making it work with relative imports. Any suggestions / pointers @patrickvonplaten?

I also didn't see `importlib` used anywhere else for similar purposes so I feel a little uneasy about bringing in additional machinery to the lib :)

yjernite commented:
> This is great! That cleans up the `modeling_(tf_)utils` a lot!
> 
> I'm thinking that `modeling_generation_utils.py` and `modeling_tf_generation_utils.py` would probably be better names. When I see this I'm thinking that `generation` is another model, rather than a utility file.
> 
> Pinging @thomwolf so he can give his opinion on introducing this mixin.

Changed the names :)

thomwolf commented on src/transformers/modeling_generation_utils.py:
```diff
@@ -0,0 +1,989 @@
+logger = logging.getLogger(__name__)
+
+
+class GenerationMixin:
```
I like the `Mixin`, the scope is well defined, it's great.

I think I would also rename the files just `generation_utils.py` and `generation_tf_utils.py`

yjernite commented:
I moved `shape_list` back to the main `modeling_tf_utils.py` (duplicated in 'generation_tf_utils.py') and renamed the files to Thom's suggestion. Should be ready to merge @LysandreJik !
